### PR TITLE
[PWN-5957] Create coordinators, where it's needed

### PR DESCRIPTION
**Pull request by qwerkly**
## Link jira to issue
https://p2pvalidator.atlassian.net/browse/PWN-5957

## Description of the changes
-

### CODE REVIEW

File: p2p_wallet/Coordinators/AppCoordinator/AppCoordinator.swift [hunk 1]
```diff
@@ -88,15 +88,25 @@ class AppCoordinator: Coordinator<Void> {
...
         }
 
-        let vm = Main.ViewModel()
-        let vc = Main.ViewController(viewModel: vm)
-        vc.authenticateWhenAppears = showAuthenticationOnMainOnAppear
-        hideLoadingAndTransitionTo(vc)
+        let coordinator = TabBarCoordinator(window: window, authenticateWhenAppears: showAuthenticationOnMainOnAppear)
+        coordinate(to: coordinator)
+            .sink(receiveValue: {
+                debugPrint("loool")
```
Comment (sidorov-panda) [inline p2p_wallet/Coordinators/AppCoordinator/AppCoordinator.swift:100]: remove?

File: p2p_wallet/Coordinators/AppCoordinator/AppCoordinator.swift [hunk 2]
```diff
@@ -88,15 +88,25 @@ class AppCoordinator: Coordinator<Void> {
...
+            })
+            .store(in: &subscriptions)
+
+//        let coordinator = MainCoordinator(window: window, authenticateWhenAppears: showAuthenticationOnMainOnAppear)
```
Comment (sidorov-panda) [inline p2p_wallet/Coordinators/AppCoordinator/AppCoordinator.swift:104]: remove?

File: p2p_wallet/Coordinators/SwapCoordinator.swift [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        view.doneHandler = { [weak self] in
+            self?.navigationController.popToRootViewController(animated: true)
+            self?.subject.send(.done)
+        }
+        view.onClose = { [weak self] in
+            self?.subject.send(.cancel)
+        }
+        navigationController.show(view, sender: nil)
+
+        return subject.eraseToAnyPublisher()
```
Comment (sidorov-panda) [inline p2p_wallet/Coordinators/SwapCoordinator.swift:45]: prefix(1)

File: p2p_wallet/Coordinators/SendCoordinator.swift [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+        sendCoordinator?.doneHandler = { [weak self] in
+            self?.navigationController.popToRootViewController(animated: true)
+            self?.subject.send(.done)
+        }
+        let view = sendCoordinator?.start(hidesBottomBarWhenPushed: true)
+        view?.onClose = { [weak self] in
+            self?.subject.send(.cancel)
+        }
+
+        return subject.eraseToAnyPublisher()
```
Comment (sidorov-panda) [inline p2p_wallet/Coordinators/SendCoordinator.swift:51]: prefix(1)

File: p2p_wallet/Coordinators/TabBar/TabBarCoordinator.swift [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+        tabBarController = TabBarController(
+            viewModel: TabBarViewModel(),
+            authenticateWhenAppears: authenticateWhenAppears
+        )
+        super.init()
+        listenMiddleButton()
+    }
+
+    deinit {
+        debugPrint("--- TabBarCoordinator deinit")
```
Comment (sidorov-panda) [inline p2p_wallet/Coordinators/TabBar/TabBarCoordinator.swift:33]: remove?

File: p2p_wallet/Coordinators/WalletDetailCoordinator.swift [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+            self?.navigationController.popToRootViewController(animated: true)
+            self?.subject.send(.done)
+        }
+        view.onClose = { [weak self] in
+            self?.subject.send(.cancel)
+        }
+
+        navigationController.show(view, sender: nil)
+
+        return subject.eraseToAnyPublisher()
```
Comment (sidorov-panda) [inline p2p_wallet/Coordinators/WalletDetailCoordinator.swift:45]: prefix(1)

**Review by sidorov-panda — CHANGES_REQUESTED** [review]

File: p2p_wallet/Coordinators/TabBar/HomeCoordinator.swift [hunk 7]
```diff
@@ -155,44 +160,51 @@ final class HomeCoordinator: Coordinator<Void> {
             .store(in: &subscriptions)
         tokensViewModel.sendShow
             .sink(receiveValue: { [unowned self, weak tokensViewModel] in
-                Task {
-                    do {
-                        let done = await sendToken()
-                        if done {
+                let coordinator = SendCoordinator(navigationController: navigationController, pubKey: nil)
```
Comment (sidorov-panda) [inline p2p_wallet/Coordinators/TabBar/HomeCoordinator.swift:163]: flatMap

File: p2p_wallet/Coordinators/TabBar/HomeCoordinator.swift [hunk 8]
```diff
@@ -155,44 +160,51 @@ final class HomeCoordinator: Coordinator<Void> {
...
+                    .store(in: &subscriptions)
             })
             .store(in: &subscriptions)
         tokensViewModel.swapShow
             .sink(receiveValue: { [unowned self, weak tokensViewModel] in
-                Task {
-                    do {
-                        let done = await showSwap()
-                        if done {
+                let coordinator = SwapCoordinator(navigationController: navigationController, initialWallet: nil)
```
Comment (sidorov-panda) [inline p2p_wallet/Coordinators/TabBar/HomeCoordinator.swift:178]: flatMap

File: p2p_wallet/Scenes/TabBar/TabBarController.swift [hunk 9]
```diff
@@ -5,39 +5,66 @@
...
+    }
+
+    @available(*, unavailable)
+    public required init?(coder _: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+
+    deinit {
+        onClose?()
+        print("lololool")
```
Comment (sidorov-panda) [inline p2p_wallet/Scenes/TabBar/TabBarController.swift:52]: erase?

**Review by sidorov-panda — CHANGES_REQUESTED** [review]

**Review by bigearsenal — DISMISSED** [review]

**Comment by bigearsenal** [comment]
Dupplicated https://github.com/p2p-org/p2p-wallet-ios/pull/823
Will merge with [823](https://github.com/p2p-org/p2p-wallet-ios/pull/823)

**Pull request closed without merging**